### ISSUE TITLE
yz/ibrowse configuration and interaction with Solr

### ISSUE THREAD

**Issue by wbrown** [basho/yokozuna#330]
Currentlly, Yokozuna uses the `ibrowse` application to stage index updates to Solr.  The default `ibrowse` settings are decidedly non-optimal with a default per host:port pair maximum of 10 sessions and 10 pipeline depth.

As the Solr indexes grow to millions of entries, the latency of the index updates exerts a backpressure upon Riak, slowing down key updates and PUTs.  `htop` shows a back and forth pattern of Riak and Solr exchanging CPU cycles.  At this point, Riak performance completely tanks, especially with the LevelDB backend.

Working with @evanmcc -- we found that the `ibrowse` author doesn't like to do configuration like the rest of the Erlang world, unfortunately.  He suggested a configuration line update, that I had to alter -- hooray for documentation that doesn't match reality.

`${RIAK_HOME}/lib/ibrowse-4.0.1/priv/ibrowse.conf` had the following line added:

```
{dest, "localhost", 8093, 100, 100000, []}.
```

Note that `{127,0,0,1}` in place of `localhost` did not work.  It really wants the string hostname, probably because `yz` is using `localhost`.  By the way, if you do this, this will default to IPV6 for localhost on some systems, which is an interesting and potentially problematic side effect.

The above settings set the connection maximum to `100` with a pipeline depth of `10000`.  I restarted Riak, and sure enough, I got that amount of connections which is an indicator that I may need to raise the limit higher for my throughput of about 400-500 keys indexed a second.

```
[root@altostratus log]#  netstat -anp | grep 8093 | grep EST | wc -l200
200
```

Checking for `TIME_WAIT` revealed hundreds of connections left over, which is indicative that there is no pipelining. 

```
[root@altostratus log]#  netstat -anp | grep 8093 | grep TIME_WAIT | wc -l
594
```

@coderoshi confirmed my suspicion.  This is problematic for deployments that index hundreds or thousands of keys a second with the overhead of establishing a socket connection for every index update.

So, I'd suggest the following changes:
- Add a more reasonable default max session value for `localhost:8093`
- Use HTTP pipelining so that we don't run out of socket descriptors/file handles and also so we don't incur the punishing latency and overhead of a TCP connection for every key update.

**Comment by rzezeski** [comment]
@wbrown Thank you very much for this informative issue. I always had a suspicion the current ibrowse setup was not optimal but I have yet to analyze it as closely as you have. Things are really busy right now but I hope someone might be able to dig into this issue in the coming weeks. Feel free to submit your own patch if you feel inclined to.

**Comment by wbrown** [comment]
@rzezeski I'd be happy to submit a patch for the configuration file.  It may not be optimal for other use cases though.  Writing HTTP pipelining into the Yokozuna interface to Solr is probably a bit beyond my current time budget though, as I'd have to learn more Erlang. :)

**Comment by rzezeski** [comment]
@wbrown Yea in this case I would want it configurable via our standard configuration. Which would require editing the configuration schema file and passing to ibrowse. But it sounds like you and @evanmcc discovered passing the pool and pipeline size programmatically doesn't work. Perhaps a patch to ibrowse is also in order. I'm not sure, I need more info and need to look at the ibrowse code.

**Comment by rzezeski** [comment]
@wbrown 

A quick look at ibrowse suggests that it automatically does the pipelining. I'm not sure the excess `TIME_WAIT` is necessarily indicative of requests not being pipelined. The local Solr process will have connections originating from two different sources:
1. the localhost when a user query comes in on that node;
2. a remote host when a Solr distributed search is being executed.

I wonder if it is case 2 that is causing those excess connections to be created. Can you verify with netstat that all `TIME_WAIT` on 8093 are localhost => localhost.

**Comment by wbrown** [comment]
@rzezeski Yes, ibrowse does do the pipelining.  However, this does not appear to be in effect, as Yokozuna appears to be doing each update request individually, and causing a new socket connect via ibrowse.

There are no searches going on while I am doing this testing -- all connections to Solr are coming via localhost.  The higher I raise the socket limit for ibrowse, the more the lingering TIME_WAITs increase.

**Comment by rzezeski** [comment]
Oh, could it be because, according to RFC 2616, you are not supposed to pipeline non-idempotent requests? Yokozuna uses `POST` to send the update.

http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html

That said, keep alive should be in play and those connections should be re-used. The requests just wouldn't be pipelined. I'm not sure why it would not reuse existing conns. Perhaps there is a bug with ibrowse connection reuse.

Another thing to note is that the jetty config (`lib/yokozuna/priv/solr/etc/jetty.xml`) has the connection thread pool set to 10,000 max threads. Just for laughs I am curious what happens if you set that equal to ibrowse connection pool size and set the `lowResourceMaxIdleTime` equal to the `maxIdleTime` on the socket connector.

**Comment by wbrown** [comment]
Setting up for this test -- setting the connection thread pool to 150 max threads.  I don't see any configuration option or default in ibrowse that tells me what the maxIdleTime would be though.

**Comment by wbrown** [comment]
Nevermind -- found the parameter you're asking for, in Solr.  Here goes.

**Comment by rzezeski** [comment]
> I don't see any configuration option or default in ibrowse that tells me what the maxIdleTime would be though.

That's a good point. I'm not sure what the defaults are for connection evictions in ibrowse.

**Comment by wbrown** [comment]
Wiith ibrowse set to allow up to 150 connections, we get 150 connections to Solr as shown below.

```
[root@altostratus ~]# netstat -anp | grep :8093 | grep EST | wc -l
300
```

We get lots of `TIME_WAIT`s still:

```
[root@altostratus ~]# netstat -anp | grep :8093 | grep TIME_WAIT | wc -l
2140
```

One parameter that I changed that seems to have helped was increasing Solr's `Acceptors` value to `10`.   Which would make sense, as the default is `2`, and if it's busy handling TCP handshakes, Riak would get backed up.  But it does make the `TIME_WAIT`s apparently worse, which makes sense as well.

To sum up, the pertinent values are:
- `maxThreads` to `150`
- `Acceptors` to `10`
- `lowResourcesMaxIdleTime` to `50000`

**Comment by rzezeski** [comment]
Linking this with issue #320.

**Comment by rzezeski** [comment]
Update: @coderoshi is going to be looking into this soon.

**Comment by coderoshi** [comment]
Closing this, moving all conversations to the related PR #358

**Comment by rzezeski** [comment]
> @rzezeski Yes, ibrowse does do the pipelining. However, this does not appear to be in effect, as Yokozuna appears to be doing each update request individually, and causing a new socket connect via ibrowse.

@wbrown While researching and reviewing @coderoshi PR #358 I have
found that, at least on OmniOS, this is **NOT** the case. I've used
DTrace to confirm that connections are being reused over multiple
YZ->Solr requests. I can reproduce the `solr.log` errors you've seen
as well as various errors in Riak's `console.log` that are very
interesting re ibrowse. But I have yet to see anything even close to
your build up of `TIME_WAIT`. I am still investigating but I thought
this information was worth sharing now in case you are still wrestling
this issue. I hope to try my experiments on a Linux distro (probably
ubuntu or centos) after I understand this issue more.

**Comment by rzezeski** [comment]
@wbrown So I've determined, at least with the current Yokozuna code,
raising your pool size will actually hurt you rather than
help. ibrowse does reuse connections and pipelines but a combination
of its load balancer algorithm and default `inactivity_timeout` are
working against you.

It turns out ibrowse always wants to have the pool filled with
connections. If you set the pool to 100 then it always wants 100
connections open. When a request is made it checks the number of conns
in the pool, if less than max then it opens a new one
[[1]](https://github.com/cmullaparthi/ibrowse/blob/v4.0.1/src/ibrowse_lb.erl#L126). If the pool is full then it searched for the first
conn that doesn't have it's pipeline full, and it **ALWAYS** starts
with the first conn. Combine this with the fact that the default
`inactivity_timeout` is 10 seconds and you have a recipe for
`TIME-WAIT` loading [[2]](http://www.isi.edu/touch/pubs/infocomm99/infocomm99-web/). ibrowse will fill the pool but then
start pipelining only those on the front of the list. The conns on the
end of the list will go stale and then close due to
`inactivity_timeout`. Then ibrowse will go back to the state where the
pool isn't full and start opening fresh conns. Repeat over and over.

This is why you are seeing time-wait load and this is why setting
higher limits alone will not help.

I wrote a dtrace script to track established connections and ran some
benchmarks designed to overwhelm Solr. With a pool/pipeline size of 20
I see around 80 new connections established a minute. Increasing the
pool/pipeline to 100 and I see 230+.

pool=20, pipeline=20:

```
root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 87

root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 87

root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

10.0.1.201:8087 1
127.0.0.1:8093 81
```

pool=100, pipeline=100:

```
root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 262

root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 236

root@oldboy:~/yz358# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 244
```

I have some ideas for making this better but thought I'd let you know
now.

tcp-conn-rate.d:

```
#pragma D option quiet

dtrace:::BEGIN
{
    printf("Monitor rate of ESTABLISHED passive connections over 1 minute");
}

tcp:::accept-established
{
    @est[args[3]->tcps_laddr, args[3]->tcps_lport] = count();
}

profile:::tick-1sec
{
    printf(".");
}

profile:::tick-60sec
{
    printf("\n\n");
    printa("%s:%d %@d\n", @est);
    exit(0);
}
```

/cc @coderoshi

**Comment by wbrown** [comment]
@rzezeski Wow, that's really interesting.

And gives me a headache, as this means that someone is going to need to carefully measure the latency of the entire pipeline to figure out an optimal connection pool size.

Incidentally, I really do dig DTrace -- it's one of the great reasons to use OmniOS and SmartOS.

**Comment by rzezeski** [comment]
> And gives me a headache, as this means that someone is going to need to carefully measure the latency of the entire pipeline to figure out an optimal connection pool size.

Yes, by setting pipeline to 1 I get zero connection churn because they
are constantly used but then I get a ton of `{error, retry_later}` in
the Riak `console.log` because the pipes are full. I tried setting
`inactivity_timeout` via `ibrowse.conf` but this didn't seem to
work. I plan to play with that more.

**Comment by wbrown** [comment]
Taking a look at the `ibrowse` code, looks like there's two possible sources of values for `inactivity_timeout`.  If I'm reading the Erlang correctly, when the request itself has options passed in, it uses the `inactivity_timeout` on a per request basis.

However, in the next breath, if the request is `undefined` i.e. probably when the library options are being initialized, it reads the `inactivity_timeout` value first via `get_config_value` and then if that doesn't return an `integer`, it then looks at the environment with `get_env` for the desired value.  If the desired value is nonzero, it uses it -- if it doesn't, it sets it to the default of `10000ms`.

https://github.com/cmullaparthi/ibrowse/blob/master/src/ibrowse_http_client.erl#L1993

``` erlang
get_inac_timeout(#state{cur_req = #request{options = Opts}}) ->
    get_value(inactivity_timeout, Opts, infinity);
get_inac_timeout(#state{cur_req = undefined}) ->
    case ibrowse:get_config_value(inactivity_timeout, undefined) of
        Val when is_integer(Val) ->
            Val;
        _ ->
            case application:get_env(ibrowse, inactivity_timeout) of
                {ok, Val} when is_integer(Val), Val > 0 ->
                    Val;
                _ ->
                    10000
            end
    end.
```

So, looking at how Yokozuna uses `ibrowse`:

https://github.com/basho/yokozuna/blob/8715c5c8e8189bf8bd839ce634317908179f80ed/src/yz_solr.erl#L65

``` erlang
    Opts = [{response_format, binary}],
    case ibrowse:send_req(URL, Headers, post, JSON, Opts) of
        {ok, "200", _, _} -> ok;
        Err -> throw({"Failed to commit", Err})
    end.
```

Here, we initialize an Opts array, pass it to `ibrowse:send_req`.  Looks normal.

Digging into `ibrowse:get_config_value` we see at https://github.com/cmullaparthi/ibrowse/blob/8d9ccc7a8b1a6febb272667bee2fa2b5f38e8a75/src/ibrowse.erl#L810

``` erlang
%% @doc Internal export
get_config_value(Key, DefVal) ->
    try
        case ets:lookup(ibrowse_conf, Key) of
            [] ->
                DefVal;
            [#ibrowse_conf{value = V}] ->
                V
        end
    catch
        error:badarg ->
            throw({error, ibrowse_not_running})
    end.
```

So, it looks only in two places for the configuration of `inactivity_timeout`, `ibrowse.conf` and the parameters passed in, defaulting to `10000ms` timeout.

One observation I have to share -- the documentation doesn't match reality.  When I opened the issue, I gave a string that worked:

```
{dest, "localhost", 8093, 100, 100000, []}.
```

It would appear that if we adjust this to:

```
{dest, "localhost", 8093, 100, 100000, [{inactivity_timeout, 60000}]
```

That it should work.  What configuration string did you attempt that didn't work?  I'll go and try this out later, when I have a little more free time, and when my test nodes aren't doing dataset load. :)

**Comment by rzezeski** [comment]
> It would appear that if we adjust this to:
> 
> {dest, "localhost", 8093, 100, 100000, [{inactivity_timeout, 60000}]
> That it should work. What configuration string did you attempt that didn't work? I'll go and try this out later, when I have a little more free time, and when my test nodes aren't doing dataset load. :)

I tried exactly this and it didn't work. But you should try too to verify.
